lib/stream/cache.go

```diff
@@ -0,0 +1,198 @@
+//	if len(b.elems) == 0 {
+//		b.elems = append(b.elems, &elem)
+//		b.lowest = height
+//
+//		return
+//	}
+//	if height < start {
+//		// Add to the front.
+//		growLen := int(start) - int(height)
+//		grow := make([]*E, growLen)
+//		grow[0] = &elem
+//		b.elems = append(grow, b.elems...)
+//		b.lowest = height
+//
+//		return
+//	} else if height > end {
+//		// Add to the back.
+//		growLen := int(height) - int(end)
+//		grow := make([]*E, growLen)
+//		grow[len(grow)-1] = &elem
+//		b.elems = append(b.elems, grow...)
+//
+//		return
+//	}
+//
+//	// Add to middle if not already present
+//	index := int(height) - int(start)
+//	b.elems[index] = &elem
+//
+//	return
+// }
+//
+// func (b *sliceCache[E]) Get(height uint64) []E {
+//	b.mu.RLock()
+//	defer b.mu.RUnlock()
+//
+//	startIndex, ok := umath.Subtract(height, b.lowest)
+//	if !ok {
+//		return nil
+//	}
+//
+//	var resp []E
+//	for i := startIndex; i < uint64(len(b.elems)); i++ {
+//		elem := b.elems[i]
+//		if elem == nil {
+//			break
+//		}
+//
+//		resp = append(resp, *elem)
+//	}
+//
+//	return resp
+// }
+
+// NewMapCache returns a new map based cache using the provided limit.
+func NewMapCache[E any](limit int) Cache[E] {
+	return &mapCache[E]{
+		limit: limit,
+		elems: make(map[uint64]E),
+	}
+}
+
+// mapCache implements Cache storing elements in a map.
+// It will only retain the highest `limit` elements.
+// It assumes elements are roughly sequential and that heights are not sparse.
+type mapCache[E any] struct {
+	limit int // Immutable
+
+	// Mutable fields
+	mu     sync.RWMutex
+	elems  map[uint64]E
+	lowest uint64 // Invalid while elems is empty
+}
+
+// Get returns all strictly sequential elements in the cache from the provided height (inclusive).
+func (c *mapCache[E]) Get(from uint64) ([]E, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	var resp []E
+	for height := from; ; height++ {
+		elem, ok := c.elems[height]
+		if !ok {
+			break
+		} else if height < c.lowest {
+			return nil, errors.New("height < lowest [BUG]") // Sanity check
+		}
+
+		resp = append(resp, elem)
+	}
+
+	return resp, nil
+}
+
+// Set adds the provided elements to the cache starting from the provided height (inclusive).
+// Elements are assumed to be in strictly sequential order.
+func (c *mapCache[E]) Set(from uint64, elems []E) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for i, elem := range elems {
+		height := from + uint64(i)
+
+		if len(c.elems) >= c.limit && height < c.lowest {
+			// Don't add too low elements when cache is full, they will just be deleted below.
+			continue
+		}
+
+		if len(c.elems) == 0 {
+			c.lowest = height // Initialize lowest when first element added.
+		}
+
+		c.elems[height] = elem
+
+		if height < c.lowest {
+			c.lowest = height // Update lowest
+		}
+	}
+
+	for len(c.elems) > c.limit {
+		delete(c.elems, c.lowest)
+
+		// Find the new lowest
+		for {
+			c.lowest++
+			if _, ok := c.elems[c.lowest]; ok {
+				break
+			}
+		}
```

Comment:
Since you rely on the `lowest` to point to some element in the map eventually anyway, why do we need the second loop advancing to the next existing element? I think `delete` for a non-existing key is a noop and will do all the same checks you're doing manually right now.

Reply:
it is about correctness for me. 

My definition: c.lowest is the lowest height in the cache or math.Maxuint64 while the cache is empty.

Are you suggesting: c.lower is lower or equaled to lowest elements in the cache...?

Reply:
> c.lower is lower or equaled to lowest elements in the cache...?

Yes! This is how I read the code:

```
while having too many elements
  a. remove lowest
  b. find the next lowest by incrementing the pointer to something that is present
  c. repeat
```

So it feels like the step b. can be simply replaced by an `increment the pointer`.

Because both versions will work if the invariant holds: `c.lowest` is below `limit` elements still present in the cache.

But no strong feels, I'm just saying when reading the code and explaining correctness to myself I "optimize" this loop "away" because I can see how it adds extra asserts for anything.